src/errors.rs

```diff
@@ -0,0 +1,158 @@
+#[derive(Debug, Eq, PartialEq)]
+pub enum StakeActionError<ErrorType> {
+    StakeNotFound,
+    Error(ErrorType),
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum TransferFromAccountError {
+    InsufficientBalance,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum StakingError {
+    CannotStakeZero,
+    CannotStakeLessThanMinimumBalance,
+    AlreadyStaked,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum StakingFromAccountError {
+    StakingError(StakingError),
+    InsufficientBalanceInSourceAccount,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum IncreasingStakeError {
+    NotStaked,
+    CannotChangeStakeByZero,
+    CannotIncreaseStakeWhileUnstaking,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum IncreasingStakeFromAccountError {
+    IncreasingStakeError(IncreasingStakeError),
+    InsufficientBalanceInSourceAccount,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum DecreasingStakeError {
+    NotStaked,
+    CannotChangeStakeByZero,
+    CannotDecreaseStakeWhileOngoingSlahes,
+    InsufficientStake,
+    CannotDecreaseStakeWhileUnstaking,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum InitiateSlashingError {
+    NotStaked,
+    SlashPeriodShouldBeGreaterThanZero,
+    SlashAmountShouldBeGreaterThanZero,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum PauseSlashingError {
+    SlashNotFound,
+    NotStaked,
+    AlreadyPaused,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum ResumeSlashingError {
+    SlashNotFound,
+    NotStaked,
+    NotPaused,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum CancelSlashingError {
+    SlashNotFound,
+    NotStaked,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum UnstakingError {
+    NotStaked,
+    AlreadyUnstaking,
+    CannotUnstakeWhileSlashesOngoing,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum InitiateUnstakingError {
+    UnstakingError(UnstakingError),
+    UnstakingPeriodShouldBeGreaterThanZero,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum PauseUnstakingError {
+    NotStaked,
+    NotUnstaking,
+    AlreadyPaused,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum ResumeUnstakingError {
+    NotStaked,
+    NotUnstaking,
+    NotPaused,
+}
+
+impl<ErrorType> From<ErrorType> for StakeActionError<ErrorType> {
```

Comment:
So I think the macro approach works well when working with simple cases and where types are global. I would much rather write one From impl rather than many macros. I've found working with macros in IDEs also very bad experience.